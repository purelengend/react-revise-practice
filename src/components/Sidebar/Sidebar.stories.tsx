import type { Meta, StoryObj } from "@storybook/react";

// Components
import Sidebar from ".";

// Contexts

const meta: Meta<typeof Sidebar> = {
  title: "Components/Sidebar",
  component: Sidebar,
  decorators: [(Story) => <Story />],
  argTypes: {
    isOpen: {
      description: "Whether the sidebar is open or closed",
    },
    onClose: {
      description: "Function to close the sidebar",
    },
  },
  args: {
    isOpen: true,
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof Sidebar>;

export const Primary: Story = {};
