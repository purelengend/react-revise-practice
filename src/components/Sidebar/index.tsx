import { CloseIcon } from "@chakra-ui/icons";
import {
  Center,
  Drawer,
  DrawerCloseButton,
  DrawerContent,
  DrawerOverlay,
  HStack,
  Link,
  Text,
  VStack,
} from "@chakra-ui/react";
import { memo } from "react";
import { Link as ReactRouterLink } from "react-router-dom";

// Components
import { UserCard } from "../common";

// Constants
import { ROLES, SIDEBAR_ITEM_LIST } from "@/constants";

export type SidebarProps = {
  sidebarState: boolean;
  onCloseSidebar: () => void;
};

const Sidebar = memo(({ sidebarState, onCloseSidebar }: SidebarProps) => {
  return (
    <Drawer
      placement="left"
      isOpen={sidebarState}
      onClose={onCloseSidebar}
      closeOnOverlayClick={true}
      variant={{
        md: "clickThrough",
      }}
    >
      <DrawerOverlay bg="rgba(101, 101, 101, .6)"></DrawerOverlay>

      <DrawerContent
        bg="white.200"
        position="relative"
        maxW={270}
        px={6.25}
        py={4.5}
      >
        <Center borderLeft="4px solid" borderLeftColor="yellow.100">
          <Link
            as={ReactRouterLink}
            to="/"
            w={220}
            fontWeight="bold"
            fontSize="lg"
            textAlign="center"
            _hover={{ textDecoration: "none" }}
          >
            CRUD OPERATIONS
          </Link>
        </Center>

        <UserCard mt={13.5} name="Karthi Madesh" role={ROLES.ADMIN} />
        <VStack mt={22.5} px={3.75}>
          {SIDEBAR_ITEM_LIST.map((item) => {
            return (
              <Link
                as={ReactRouterLink}
                key={item.title}
                to={item.destination}
                _hover={{ textDecoration: "none", bg: "yellow.200" }}
                w="full"
                py={2.5}
                borderRadius={4}
              >
                <HStack pl={10} gap={4}>
                  {item.icon}
                  <Text textAlign="center" fontWeight="500" fontSize="sm">
                    {item.title}
                  </Text>
                </HStack>
              </Link>
            );
          })}
        </VStack>
        <DrawerCloseButton
          position="absolute"
          boxSize={6}
          top={0}
          right={1}
          onClick={onCloseSidebar}
        >
          <CloseIcon w={3} h={3} />
        </DrawerCloseButton>
      </DrawerContent>
    </Drawer>
  );
});

export default Sidebar;
