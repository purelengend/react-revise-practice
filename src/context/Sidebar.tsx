// Hooks
import { useSidebar } from "@/hooks";
import { createContext, useCallback } from "react";

export type SidebarContextProps = {
  sidebarState: boolean;
  openSidebar: () => void;
  closeSidebar: () => void;
};

export const SidebarContext = createContext<SidebarContextProps>(
  {} as SidebarContextProps,
);

export const SidebarContextProvider = (props: {
  children: React.ReactNode;
}) => {
  const { sidebarState, openSidebar, closeSidebar } = useSidebar();

  const openSidebarCallback = useCallback(() => {
    openSidebar;
  }, [openSidebar]);

  const closeSidebarCallback = useCallback(() => {
    closeSidebar;
  }, [closeSidebar]);

  return (
    <SidebarContext.Provider
      value={{
        sidebarState,
        openSidebar: openSidebarCallback,
        closeSidebar: closeSidebarCallback,
      }}
    >
      {props.children}
    </SidebarContext.Provider>
  );
};
