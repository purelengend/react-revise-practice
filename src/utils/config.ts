// Check import.meta.env is undefined in test environment
export const required = (key: string, defaultValue = undefined) => {
  if (import.meta.env) return import.meta.env[key];
  return defaultValue;
};
