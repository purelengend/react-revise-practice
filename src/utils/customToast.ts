import { TOAST_DEFAULT_OPTIONS } from "@/constants";

import { AlertStatus, UseToastOptions } from "@chakra-ui/react";
export const customToast = (
  title: string,
  description: string,
  status: AlertStatus,
): UseToastOptions => {
  if (status === "success") {
    return {
      ...TOAST_DEFAULT_OPTIONS.SUCCESS,
      title,
      description,
    };
  }

  return {
    ...TOAST_DEFAULT_OPTIONS.ERROR,
    title,
    description,
  };
};
