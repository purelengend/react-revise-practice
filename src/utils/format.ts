export const formatPhoneNumber = (phoneNumber: string) => {
  // Check if the phone number is valid (10 digits)
  if (!/^\d{10}$/.test(phoneNumber)) {
    throw new Error("Invalid phone number. Must be a 10-digit number.");
  }

  // Format the phone number
  const areaCode = phoneNumber.slice(0, 3);
  const middlePart = phoneNumber.slice(3, 6);
  const lastPart = phoneNumber.slice(6, 10);

  return `${areaCode}-${middlePart}-${lastPart}`;
};
