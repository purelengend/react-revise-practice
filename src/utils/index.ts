export * from "./test-utils";
export * from "./config";
export * from "./customToast";
