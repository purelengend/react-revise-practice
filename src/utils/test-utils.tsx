/* eslint-disable react-refresh/only-export-components */
import { RenderOptions, render } from "@testing-library/react";
import { ReactElement, ReactNode } from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { MemoryRouter } from "react-router-dom";

const queryClient = new QueryClient({});
export const AllTheProviders = (props: { children: ReactNode }) => {
  return (
    <MemoryRouter initialEntries={["/"]}>
      <QueryClientProvider client={queryClient}>
        {props.children}
      </QueryClientProvider>
    </MemoryRouter>
  );
};

export const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
