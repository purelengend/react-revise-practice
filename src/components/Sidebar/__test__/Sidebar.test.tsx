import { act } from "react";
import Sidebar from "..";
import { customRender } from "@/utils";

describe("Sidebar test cases", () => {
  const mockOnClose = jest.fn();

  const setup = () =>
    customRender(<Sidebar isOpen={true} onClose={mockOnClose} />);

  it("should render correctly", async () => {
    const { container } = await act(() => setup());

    expect(container).toMatchSnapshot();
  });
});
