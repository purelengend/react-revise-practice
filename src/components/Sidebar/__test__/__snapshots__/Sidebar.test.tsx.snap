// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar test cases should render correctly 1`] = `<div />`;
